Use after free ASAN error in http2 adaptor

```
>           raise RuntimeError("Errors during teardown: \n\n%s" % "\n\n".join([str(e) for e in errors])) from errors[0]
E           RuntimeError: Errors during teardown: 
E           
E           Process 5771 (name=http2-test-router-a) error: returned error code 1
E           skrouterd -c http2-test-router-a.conf -I /home/runner/work/skupper-router/skupper-router/skupper-router/python
E           /__w/skupper-router/skupper-router/skupper-router/build/tests/system_test.dir/tests/system_tests_http2/Http2TestTwoRouter/setUpClass/http2-test-router-a-29.cmd
E           >>>>
E           =================================================================
E           ==5771==ERROR: AddressSanitizer: use-after-poison on address 0x61600012a620 at pc 0x00000050f43a bp 0x7fc64d272a00 sp 0x7fc64d2729f8
E           WRITE of size 8 at 0x61600012a620 thread T4
E               #0 0x50f439 in qdr_delivery_set_context /home/runner/work/skupper-router/skupper-router/skupper-router/src/router_core/delivery.c:39
E               #1 0x50f439 in on_frame_recv_callback /home/runner/work/skupper-router/skupper-router/skupper-router/src/adaptors/http2/http2_adaptor.c:1255
E               #2 0x7fc656767004 in nghttp2_session_mem_recv (/lib64/libnghttp2.so.14+0xf004)
E               #3 0x4ef753 in push_rx_buffer_to_nghttp2 /home/runner/work/skupper-router/skupper-router/skupper-router/src/adaptors/http2/http2_adaptor.c:2593
E               #4 0x4f99f1 in handle_incoming_http /home/runner/work/skupper-router/skupper-router/skupper-router/src/adaptors/http2/http2_adaptor.c:2720
E               #5 0x51db2c in handle_connection_event /home/runner/work/skupper-router/skupper-router/skupper-router/src/adaptors/http2/http2_adaptor.c:3277
E               #6 0x8175dd in handle_event_with_context /home/runner/work/skupper-router/skupper-router/skupper-router/src/server.c:805
E               #7 0x8175dd in do_handle_raw_connection_event /home/runner/work/skupper-router/skupper-router/skupper-router/src/server.c:811
E               #8 0x8175dd in handle /home/runner/work/skupper-router/skupper-router/skupper-router/src/server.c:1086
E               #9 0x81d775 in thread_run /home/runner/work/skupper-router/skupper-router/skupper-router/src/server.c:1118
E               #10 0x7fc6559c412c in start_thread (/lib64/libc.so.6+0x8b12c)
E               #11 0x7fc655a45bbf in __clone3 (/lib64/libc.so.6+0x10cbbf)
E           
E           0x61600012a620 is located 160 bytes inside of 512-byte region [0x61600012a580,0x61600012a780)
E           allocated by thread T4 here:
E               #0 0x7fc656fd9cdd in __interceptor_posix_memalign (/lib64/libasan.so.8+0xb9cdd)
E               #1 0x5561e1 in qd_alloc /home/runner/work/skupper-router/skupper-router/skupper-router/src/alloc_pool.c:404
E               #2 0x7716a5 in new_qdr_delivery_t /home/runner/work/skupper-router/skupper-router/skupper-router/src/router_core/delivery.c:24
E               #3 0x7716a5 in qdr_link_deliver_to /home/runner/work/skupper-router/skupper-router/skupper-router/src/router_core/transfer.c:81
E               #4 0x7fcb8c in AMQP_rx_handler /home/runner/work/skupper-router/skupper-router/skupper-router/src/router_node.c:738
E               #5 0x58c1b1 in do_receive /home/runner/work/skupper-router/skupper-router/skupper-router/src/container.c:218
E               #6 0x58c1b1 in qd_container_handle_event /home/runner/work/skupper-router/skupper-router/skupper-router/src/container.c:729
E               #7 0x817531 in handle /home/runner/work/skupper-router/skupper-router/skupper-router/src/server.c:1093
E               #8 0x81d6b0 in thread_run /home/runner/work/skupper-router/skupper-router/skupper-router/src/server.c:1118
E               #9 0x7fc6559c412c in start_thread (/lib64/libc.so.6+0x8b12c)
E           
E           Thread T4 created by T0 here:
E               #0 0x7fc656f6b3e6 in __interceptor_pthread_create (/lib64/libasan.so.8+0x4b3e6)
E               #1 0x609e9c in sys_thread /home/runner/work/skupper-router/skupper-router/skupper-router/src/posix/threading.c:167
E               #2 0x81f585 in qd_server_run /home/runner/work/skupper-router/skupper-router/skupper-router/src/server.c:1506
E               #3 0x44a6e6 in main_process /home/runner/work/skupper-router/skupper-router/skupper-router/router/src/main.c:111
E               #4 0x43b008 in main /home/runner/work/skupper-router/skupper-router/skupper-router/router/src/main.c:365
E               #5 0x7fc65596050f in __libc_start_call_main (/lib64/libc.so.6+0x2750f)
E           
E           SUMMARY: AddressSanitizer: use-after-poison /home/runner/work/skupper-router/skupper-router/skupper-router/src/router_core/delivery.c:39 in qdr_delivery_set_context
E           Shadow bytes around the buggy address:
E             0x0c2c8001d470: f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7
E             0x0c2c8001d480: f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 00 00 00
E             0x0c2c8001d490: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
E             0x0c2c8001d4a0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
E             0x0c2c8001d4b0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
E           =>0x0c2c8001d4c0: 00 00 f7 f7[f7]f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7
E             0x0c2c8001d4d0: f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7
E             0x0c2c8001d4e0: f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 00 00 00
E             0x0c2c8001d4f0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
E             0x0c2c8001d500: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
E             0x0c2c8001d510: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
E           Shadow byte legend (one shadow byte represents 8 application bytes):
E             Addressable:           00
E             Partially addressable: 01 02 03 04 05 06 07 
E             Heap left redzone:       fa
E             Freed heap region:       fd
E             Stack left redzone:      f1
E             Stack mid redzone:       f2
E             Stack right redzone:     f3
E             Stack after return:      f5
E             Stack use after scope:   f8
E             Global redzone:          f9
E             Global init order:       f6
E             Poisoned by user:        f7
E             Container overflow:      fc
E             Array cookie:            ac
E             Intra object redzone:    bb
E             ASan internal:           fe
E             Left alloca redzone:     ca
E             Right alloca redzone:    cb
E           ==5771==ABORTING
E           <<<<

/home/runner/work/skupper-router/skupper-router/skupper-router/tests/system_test.py:1209: RuntimeError
---------------------------------------------------------------------- Captured log call -----------------------------------------------------------------------
INFO     proton:_reactor.py:1053 Connecting to Url('amqp://0.0.0.0:23222')...
INFO     proton:_reactor.py:1107 Disconnected, reconnect disabled
-------------------------------------------------------------------- Captured log teardown ---------------------------------------------------------------------
INFO     proton:_reactor.py:1105 Disconnected, already closed
INFO     proton:_reactor.py:1105 Disconnected, already closed
=========================================================================== FAILURES ===========================================================================
______________________________________________________ Http2TestTwoRouter.test_zzz_http_connector_delete _______________________________________________________

self = <tests.system_tests_http2.Http2TestTwoRouter testMethod=test_zzz_http_connector_delete>

    @unittest.skipIf(skip_test(), "Python 3.7 or greater, Quart 0.13.0 or greater and curl needed to run http2 tests")
    def test_zzz_http_connector_delete(self):
>       self.check_connector_delete(client_addr=self.router_qdra.http_addresses[0],
                                    server_addr=self.router_qdrb.addresses[0],
                                    server_port=self.server_port,
                                    listener_addr=self.router_qdra.addresses[0])

/home/runner/work/skupper-router/skupper-router/skupper-router/tests/system_tests_http2.py:785: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/home/runner/work/skupper-router/skupper-router/skupper-router/tests/system_tests_http2.py:377: in check_connector_delete
    wait_http_listeners_up(listener_addr)
/home/runner/work/skupper-router/skupper-router/skupper-router/tests/http1_tests.py:529: in wait_http_listeners_up
    return _wait_adaptor_listeners_oper_status(LISTENER_TYPE, mgmt_address,
/home/runner/work/skupper-router/skupper-router/skupper-router/tests/http1_tests.py:566: in _wait_adaptor_listeners_oper_status
    mgmt = Node.connect(mgmt_address, timeout=timeout)
/home/runner/work/skupper-router/skupper-router/skupper-router/python/skupper_router/management/client.py:109: in connect
    connection = BlockingConnection(url,
/usr/local/lib64/python3.11/site-packages/proton/_utils.py:394: in __init__
    self.wait(lambda: not (self.conn.state & Endpoint.REMOTE_UNINIT),
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <proton._utils.BlockingConnection object at 0x7f765e88fa50>, condition = <function BlockingConnection.__init__.<locals>.<lambda> at 0x7f765e8a0ae0>
timeout = 300.0, msg = 'Opening connection'

    def wait(
            self,
            condition: Callable[[], bool],
            timeout: Union[None, Literal[False], float] = False,
            msg: Optional[str] = None
    ) -> None:
        """
        Process events until ``condition()`` returns ``True``.
    
        :param condition: Condition which determines when the wait will end.
        :param timeout: Timeout in seconds. If ``False``, the value of ``timeout`` used in the
            constructor of this object will be used. If ``None``, there is no timeout. Any other
            value is treated as a timeout in seconds.
        :param msg: Context message for :class:`proton.Timeout` exception
        """
        if timeout is False:
            timeout = self.timeout
        if timeout is None:
            while not condition() and not self.disconnected:
                self.container.process()
        else:
            container_timeout = self.container.timeout
            self.container.timeout = timeout
            try:
                deadline = time.time() + timeout
                while not condition() and not self.disconnected:
                    self.container.process()
                    if deadline < time.time():
                        txt = "Connection %s timed out" % self.url
                        if msg:
                            txt += ": " + msg
                        raise Timeout(txt)
            finally:
                self.container.timeout = container_timeout
        if self.disconnected and not self._is_closed():
>           raise ConnectionException(
                "Connection %s disconnected: %s" % (self.url, self.disconnected))
E           proton._exceptions.ConnectionException: Connection amqp://0.0.0.0:23222 disconnected: Condition('proton.pythonio', 'Connection refused to all addresses')

/usr/local/lib64/python3.11/site-packages/proton/_utils.py:533: ConnectionException
---------------------------------------------------------------------- Captured log call -----------------------------------------------------------------------
INFO     proton:_reactor.py:1053 Connecting to Url('amqp://0.0.0.0:23222')...
INFO     proton:_reactor.py:1107 Disconnected, reconnect disabled
---------------------- generated xml file: /__w/skupper-router/skupper-router/skupper-router/build/tests/junitxmls/system_tests_http2.xml ----------------------
=================================================================== short test summary info ====================================================================
FAILED ::Http2TestTwoRouter::test_zzz_http_connector_delete - proton._exceptions.ConnectionException: Connection amqp://0.0.0.0:23222 disconnected: Condition('proton.pythonio', 'Connection refused to all addresses')
ERROR ::Http2TestTwoRouter::test_zzz_http_connector_delete - RuntimeError: Errors during teardown: 

Process 5771 (name=http2-test-router-a) error: returned error code 1
skrouterd -c http2-test-router-a.conf -I /home/runner/work/skupper-router/skupper-router/skupper-router/python
/__w/skupper-router/skupper-router/skupper-router/build/tests/system_test.dir/tests/system_tests_http2/Http2TestTwoRouter/setUpClass/http2-test-router-a-29.cmd
>>>>
=================================================================
==5771==ERROR: AddressSanitizer: use-after-poison on address 0x61600012a620 at pc 0x00000050f43a bp 0x7fc64d272a00 sp 0x7fc64d2729f8
WRITE of size 8 at 0x61600012a620 thread T4
    #0 0x50f439 in qdr_delivery_set_context /home/runner/work/skupper-router/skupper-router/skupper-router/src/router_core/delivery.c:39
    #1 0x50f439 in on_frame_recv_callback /home/runner/work/skupper-router/skupper-router/skupper-router/src/adaptors/http2/http2_adaptor.c:1255
    #2 0x7fc656767004 in nghttp2_session_mem_recv (/lib64/libnghttp2.so.14+0xf004)
    #3 0x4ef753 in push_rx_buffer_to_nghttp2 /home/runner/work/skupper-router/skupper-router/skupper-router/src/adaptors/http2/http2_adaptor.c:2593
    #4 0x4f99f1 in handle_incoming_http /home/runner/work/skupper-router/skupper-router/skupper-router/src/adaptors/http2/http2_adaptor.c:2720
    #5 0x51db2c in handle_connection_event /home/runner/work/skupper-router/skupper-router/skupper-router/src/adaptors/http2/http2_adaptor.c:3277
    #6 0x8175dd in handle_event_with_context /home/runner/work/skupper-router/skupper-router/skupper-router/src/server.c:805
    #7 0x8175dd in do_handle_raw_connection_event /home/runner/work/skupper-router/skupper-router/skupper-router/src/server.c:811
    #8 0x8175dd in handle /home/runner/work/skupper-router/skupper-router/skupper-router/src/server.c:1086
    #9 0x81d775 in thread_run /home/runner/work/skupper-router/skupper-router/skupper-router/src/server.c:1118
    #10 0x7fc6559c412c in start_thread (/lib64/libc.so.6+0x8b12c)
    #11 0x7fc655a45bbf in __clone3 (/lib64/libc.so.6+0x10cbbf)

0x61600012a620 is located 160 bytes inside of 512-byte region [0x61600012a580,0x61600012a780)
allocated by thread T4 here:
    #0 0x7fc656fd9cdd in __interceptor_posix_memalign (/lib64/libasan.so.8+0xb9cdd)
    #1 0x5561e1 in qd_alloc /home/runner/work/skupper-router/skupper-router/skupper-router/src/alloc_pool.c:404
    #2 0x7716a5 in new_qdr_delivery_t /home/runner/work/skupper-router/skupper-router/skupper-router/src/router_core/delivery.c:24
    #3 0x7716a5 in qdr_link_deliver_to /home/runner/work/skupper-router/skupper-router/skupper-router/src/router_core/transfer.c:81
    #4 0x7fcb8c in AMQP_rx_handler /home/runner/work/skupper-router/skupper-router/skupper-router/src/router_node.c:738
    #5 0x58c1b1 in do_receive /home/runner/work/skupper-router/skupper-router/skupper-router/src/container.c:218
    #6 0x58c1b1 in qd_container_handle_event /home/runner/work/skupper-router/skupper-router/skupper-router/src/container.c:729
    #7 0x817531 in handle /home/runner/work/skupper-router/skupper-router/skupper-router/src/server.c:1093
    #8 0x81d6b0 in thread_run /home/runner/work/skupper-router/skupper-router/skupper-router/src/server.c:1118
    #9 0x7fc6559c412c in start_thread (/lib64/libc.so.6+0x8b12c)

Thread T4 created by T0 here:
    #0 0x7fc656f6b3e6 in __interceptor_pthread_create (/lib64/libasan.so.8+0x4b3e6)
    #1 0x609e9c in sys_thread /home/runner/work/skupper-router/skupper-router/skupper-router/src/posix/threading.c:167
    #2 0x81f585 in qd_server_run /home/runner/work/skupper-router/skupper-router/skupper-router/src/server.c:1506
    #3 0x44a6e6 in main_process /home/runner/work/skupper-router/skupper-router/skupper-router/router/src/main.c:111
    #4 0x43b008 in main /home/runner/work/skupper-router/skupper-router/skupper-router/router/src/main.c:365
    #5 0x7fc65596050f in __libc_start_call_main (/lib64/libc.so.6+0x2750f)

SUMMARY: AddressSanitizer: use-after-poison /home/runner/work/skupper-router/skupper-router/skupper-router/src/router_core/delivery.c:39 in qdr_delivery_set_context
Shadow bytes around the buggy address:
  0x0c2c8001d470: f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7
  0x0c2c8001d480: f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 00 00 00
  0x0c2c8001d490: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c2c8001d4a0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c2c8001d4b0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x0c2c8001d4c0: 00 00 f7 f7[f7]f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7
  0x0c2c8001d4d0: f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7
  0x0c2c8001d4e0: f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7 00 00 00
  0x0c2c8001d4f0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c2c8001d500: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c2c8001d510: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==5771==ABORTING
<<<<
====================================================== 1 failed, 95 passed, 1 error in 696.59s (0:11:36) =======================================================


97% tests passed, 1 tests failed out of 34

Total Test time (real) = 802.20 sec

The following tests FAILED:
	 56 - system_tests_http2 (Failed)
Error: Process completed with exit code 8.

```

https://github.com/skupperproject/skupper-router/actions/runs/5144827039/jobs/9261614318#step:34:4456

--- a/src/adaptors/http2/http2_adaptor.c
+++ b/src/adaptors/http2/http2_adaptor.c
@@ -382,6 +382,7 @@ static void free_http2_stream_data(qdr_http2_stream_data_t *stream_data, bool on
         qd_message_stream_data_release(stream_data->next_stream_data);
         stream_data->next_stream_data = 0;
 
+        set_stream_data_delivery_flags(stream_data, stream_data->in_dlv);
         qdr_delivery_decref(http2_adaptor->core, stream_data->in_dlv, "HTTP2 adaptor in_dlv - free_http2_stream_data");
     }
 
@@ -393,6 +394,7 @@ static void free_http2_stream_data(qdr_http2_stream_data_t *stream_data, bool on
 
         qd_message_stream_data_release(stream_data->next_stream_data);
         stream_data->next_stream_data = 0;
+        set_stream_data_delivery_flags(stream_data, stream_data->in_dlv);
         qdr_delivery_decref(http2_adaptor->core, stream_data->out_dlv,
                             "HTTP2 adaptor out_dlv - free_http2_stream_data");
     }
@@ -1251,8 +1253,17 @@ static int on_frame_recv_callback(nghttp2_session *session,
                     qdr_delivery_remote_state_updated(http2_adaptor->core, stream_data->out_dlv,
                                                       stream_data->out_dlv_local_disposition, true, 0, false);
                 }
-                if (stream_data->in_dlv) {
-                    qdr_delivery_set_context(stream_data->in_dlv, 0);
+                if (stream_data->in_dlv && !stream_data->in_dlv_decrefed) {
+                    // The stream_data->in_dlv could sometimes be freed from underneath when there
+                    // is a race between the connection close and the handling of the RST_STREAM
+                    // Procced to set the context only if the stream_data->in_dlv_decrefed has not already been
+                    // decrefed.
+                    // Fix for https://github.com/skupperproject/skupper-router/issues/1106
+                    if (stream_data->in_dlv_decrefed) {
+                        stream_data->in_dlv = 0;
+                    } else {
+                        qdr_delivery_set_context(stream_data->in_dlv, 0);
+                    }
                 }
                 qd_log(LOG_HTTP_ADAPTOR, QD_LOG_TRACE,
                        "[C%" PRIu64 "][S%" PRId32 "] HTTP2 NGHTTP2_RST_STREAM frame received, freeing stream data",
@@ -1980,8 +1991,8 @@ static void qdr_http_delivery_update(void *context, qdr_delivery_t *dlv, uint64_
                    stream_data->conn->conn_id, stream_data->stream_id);
         }
 
-        qdr_delivery_decref(http2_adaptor->core, dlv, "HTTP2 adaptor  - qdr_http_delivery_update");
         set_stream_data_delivery_flags(stream_data, dlv);
+        qdr_delivery_decref(http2_adaptor->core, dlv, "HTTP2 adaptor  - qdr_http_delivery_update");
 
         if (send_complete && stream_data->status == QD_STREAM_FULLY_CLOSED) {
             // When all the necessary HTTP2 frames have been sent and the stream is fully closed, free the stream.
@@ -2464,8 +2475,10 @@ uint64_t handle_outgoing_http(qdr_http2_stream_data_t *stream_data)
                        "] In handle_outgoing_http, qdr_delivery_remote_state_updated(stream_data->out_dlv)",
                        conn->conn_id, stream_data->stream_id);
                 stream_data->disp_updated = true;
-                qdr_delivery_decref(http2_adaptor->core, stream_data->out_dlv, "HTTP2 adaptor out_dlv - handle_outgoing_http");
                 set_stream_data_delivery_flags(stream_data, stream_data->out_dlv);
+                qdr_delivery_decref(http2_adaptor->core, stream_data->out_dlv,
+                                    "HTTP2 adaptor out_dlv - handle_outgoing_http");
+                stream_data->out_dlv = 0;
             }
         }
         qd_log(LOG_HTTP_ADAPTOR, QD_LOG_TRACE, "[C%" PRIu64 "] Finished handle_outgoing_http", conn->conn_id);

